File: ideviewer/detectors/vscode.py
"""
VS Code, Cursor, and VSCodium detector.

These IDEs share similar extension formats based on VS Code.
"""

import json
import os
import sys
from pathlib import Path
from typing import List, Optional, Dict, Any
from datetime import datetime

# Handle both direct execution and package import
try:
    from .base import BaseDetector
    from ..models import IDE, IDEType, Extension, Permission
except ImportError:
    sys.path.insert(0, str(Path(__file__).parent.parent.parent))
    from ideviewer.detectors.base import BaseDetector
    from ideviewer.models import IDE, IDEType, Extension, Permission


# Known dangerous VS Code extension permissions/capabilities
DANGEROUS_PERMISSIONS = {
    "onFileSystem": "Full file system access",
    "onStartupFinished": "Runs on startup",
    "onUri": "Can register URI handlers",
    "onAuthenticationRequest": "Authentication access",
    "onDebug": "Debugger access",
    "onTerminalProfile": "Terminal access",
    "*": "Wildcard activation (runs for everything)",
}

# Capability patterns that indicate elevated permissions
DANGEROUS_CAPABILITIES = {
    "untrustedWorkspaces.supported": "Can run in untrusted workspaces",
    "virtualWorkspaces.supported": "Virtual workspace access",
}


class VSCodeDetector(BaseDetector):
    """Detector for VS Code-based editors."""
    
    # Configuration for different VS Code variants
    VARIANTS = {
        IDEType.VSCODE: {
            "name": "Visual Studio Code",
            "executables": {
                "darwin": [
                    "/Applications/Visual Studio Code.app/Contents/Resources/app/bin/code",
                    "/usr/local/bin/code",
                ],
                "linux": [
                    "/usr/bin/code",
                    "/usr/share/code/bin/code",
                    "/snap/bin/code",
                ],
                "windows": [
                    "%LOCALAPPDATA%\\Programs\\Microsoft VS Code\\Code.exe",
                    "%ProgramFiles%\\Microsoft VS Code\\Code.exe",
                ],
            },
            "extensions_paths": {
                "darwin": ["~/.vscode/extensions"],
                "linux": ["~/.vscode/extensions"],
                "windows": ["%USERPROFILE%\\.vscode\\extensions"],
            },
            "config_paths": {
                "darwin": ["~/Library/Application Support/Code"],
                "linux": ["~/.config/Code"],
                "windows": ["%APPDATA%\\Code"],
            },
            "process_names": ["Code", "code", "Code.exe"],
        },
        IDEType.CURSOR: {
            "name": "Cursor",
            "executables": {
                "darwin": [
                    "/Applications/Cursor.app/Contents/Resources/app/bin/cursor",
                    "/usr/local/bin/cursor",
                ],
                "linux": [
                    "/usr/bin/cursor",
                    "/opt/Cursor/cursor",
                    "~/.local/bin/cursor",
                ],
                "windows": [
                    "%LOCALAPPDATA%\\Programs\\Cursor\\Cursor.exe",
                    "%LOCALAPPDATA%\\cursor\\Cursor.exe",
                ],
            },
            "extensions_paths": {
                "darwin": ["~/.cursor/extensions"],
                "linux": ["~/.cursor/extensions"],
                "windows": ["%USERPROFILE%\\.cursor\\extensions"],
            },
            "config_paths": {
                "darwin": ["~/Library/Application Support/Cursor"],
                "linux": ["~/.config/Cursor"],
                "windows": ["%APPDATA%\\Cursor"],
            },
            "process_names": ["Cursor", "cursor", "Cursor.exe"],
        },
        IDEType.VSCODIUM: {
            "name": "VSCodium",
            "executables": {
                "darwin": [
                    "/Applications/VSCodium.app/Contents/Resources/app/bin/codium",
                    "/usr/local/bin/codium",
                ],
                "linux": [
                    "/usr/bin/codium",
                    "/snap/bin/codium",
                ],
                "windows": [
                    "%LOCALAPPDATA%\\Programs\\VSCodium\\VSCodium.exe",
                    "%ProgramFiles%\\VSCodium\\VSCodium.exe",
                ],
            },
            "extensions_paths": {
                "darwin": ["~/.vscode-oss/extensions"],
                "linux": ["~/.vscode-oss/extensions"],
                "windows": ["%USERPROFILE%\\.vscode-oss\\extensions"],
            },
            "config_paths": {
                "darwin": ["~/Library/Application Support/VSCodium"],
                "linux": ["~/.config/VSCodium"],
                "windows": ["%APPDATA%\\VSCodium"],
            },
            "process_names": ["VSCodium", "codium", "VSCodium.exe"],
        },
    }
    
    def detect(self) -> List[IDE]:
        """Detect all installed VS Code variants."""
        ides = []
        
        for ide_type, config in self.VARIANTS.items():
            ide = self._detect_variant(ide_type, config)
            if ide:
                # Parse extensions for this IDE
                ide.extensions = self.parse_extensions(ide)
                ides.append(ide)
        
        return ides
    
    def _detect_variant(self, ide_type: IDEType, config: Dict[str, Any]) -> Optional[IDE]:
        """Detect a specific VS Code variant."""
        # Get platform-specific paths
        platform_key = self.system
        executables = config["executables"].get(platform_key, [])
        extensions_paths = config["extensions_paths"].get(platform_key, [])
        config_paths = config["config_paths"].get(platform_key, [])
        
        # Find executable
        install_path = None
        for exe_path in executables:
            expanded = self.expand_path(exe_path)
            if expanded.exists():
                install_path = str(expanded)
                break
        
        # Also check if just looking at extension directories (IDE might be installed differently)
        extensions_path = None
        for ext_path in extensions_paths:
            expanded = self.expand_path(ext_path)
            if expanded.exists():
                extensions_path = str(expanded)
                break
        
        # If neither executable nor extensions found, IDE is not installed
        if not install_path and not extensions_path:
            return None
        
        # Find config path
        config_path = None
        for cfg_path in config_paths:
            expanded = self.expand_path(cfg_path)
            if expanded.exists():
                config_path = str(expanded)
                break
        
        # Get version
        version = None
        if install_path:
            version = self.get_version(install_path)
        
        # Check if running
        is_running = self.is_process_running(config["process_names"])
        
        return IDE(
            ide_type=ide_type,
            name=config["name"],
            version=version,
            install_path=install_path,
            config_path=config_path,
            extensions_path=extensions_path,
            is_running=is_running,
        )
    
    def parse_extensions(self, ide: IDE) -> List[Extension]:
        """Parse all extensions for a VS Code variant."""
        extensions = []
        
        if not ide.extensions_path or not os.path.isdir(ide.extensions_path):
            return extensions
        
        extensions_dir = Path(ide.extensions_path)
        
        for ext_folder in extensions_dir.iterdir():
            if not ext_folder.is_dir():
                continue
            
            # Skip .obsolete folders
            if ext_folder.name.startswith("."):
                continue
            
            extension = self._parse_extension(ext_folder, ide.ide_type)
            if extension:
                extensions.append(extension)
        
        return extensions
    
    def _parse_extension(self, ext_folder: Path, ide_type: IDEType) -> Optional[Extension]:
        """Parse a single extension from its folder."""
        package_json = ext_folder / "package.json"
        
        if not package_json.exists():
            return None
        
        try:
            with open(package_json, "r", encoding="utf-8") as f:
                manifest = json.load(f)
        except (json.JSONDecodeError, IOError):
            return None
        
        # Extract basic info
        ext_id = manifest.get("name", ext_folder.name)
        publisher = manifest.get("publisher", "")
        if publisher:
            ext_id = f"{publisher}.{ext_id}"
        
        # Parse permissions from activation events
        permissions = self._extract_permissions(manifest)
        
        # Get marketplace URL
        marketplace_url = None
        if ide_type == IDEType.VSCODE:
            marketplace_url = f"https://marketplace.visualstudio.com/items?itemName={ext_id}"
        elif ide_type == IDEType.VSCODIUM:
            marketplace_url = f"https://open-vsx.org/extension/{publisher}/{manifest.get('name', '')}"
        
        # Get repository info
        repository = manifest.get("repository")
        if isinstance(repository, dict):
            repository = repository.get("url", "")
        
        # Get last modified time
        last_updated = None
        try:
            stat = package_json.stat()
            last_updated = datetime.fromtimestamp(stat.st_mtime)
        except OSError:
            pass
        
        # Check if builtin
        is_builtin = "ms-vscode" in str(ext_folder).lower() and "builtin" in str(ext_folder).lower()
        
        return Extension(
            id=ext_id,
            name=manifest.get("displayName", manifest.get("name", ext_folder.name)),
            version=manifest.get("version", "unknown"),
            publisher=publisher,
            maintainer=manifest.get("author", {}).get("name") if isinstance(manifest.get("author"), dict) else manifest.get("author"),
            description=manifest.get("description"),
            homepage=manifest.get("homepage"),
            repository=repository,
            license=manifest.get("license"),
            install_path=str(ext_folder),
            permissions=permissions,
            contributes=self._summarize_contributes(manifest.get("contributes", {})),
            dependencies=list(manifest.get("extensionDependencies", [])),
            enabled=True,  # Would need to check settings to determine this
            builtin=is_builtin,
            last_updated=last_updated,
            marketplace_url=marketplace_url,
            activation_events=manifest.get("activationEvents", []),
            capabilities=manifest.get("capabilities", {}),
        )
    
    def _extract_permissions(self, manifest: Dict[str, Any]) -> List[Permission]:
        """Extract permissions from extension manifest."""
        permissions = []
        
        # Check activation events
        activation_events = manifest.get("activationEvents", [])
        for event in activation_events:
            event_type = event.split(":")[0] if ":" in event else event
            
            if event_type in DANGEROUS_PERMISSIONS:
                permissions.append(Permission(
                    name=event_type,
                    description=DANGEROUS_PERMISSIONS[event_type],
                    is_dangerous=True,
                ))
            elif event == "*":
                permissions.append(Permission(
                    name="*",
                    description=DANGEROUS_PERMISSIONS["*"],
                    is_dangerous=True,
                ))
        
        # Check capabilities
        capabilities = manifest.get("capabilities", {})
        for cap_name, cap_desc in DANGEROUS_CAPABILITIES.items():
            if cap_name in str(capabilities):
                permissions.append(Permission(
                    name=cap_name,
                    description=cap_desc,
                    is_dangerous=True,
                ))
        
        # Check contributes for sensitive contributions
        contributes = manifest.get("contributes", {})
        
        if "authentication" in contributes:
            permissions.append(Permission(
                name="authentication",
                description="Provides authentication providers",
                is_dangerous=True,
            ))
        
        if "terminal" in contributes:
            permissions.append(Permission(
                name="terminal",
                description="Terminal integration",
                is_dangerous=True,
            ))
        
        if "debuggers" in contributes:
            permissions.append(Permission(
                name="debuggers",
                description="Debugger integration",
                is_dangerous=False,
            ))
        
        if "taskDefinitions" in contributes:
            permissions.append(Permission(
                name="taskDefinitions",
                description="Can define tasks (may execute commands)",
                is_dangerous=True,
            ))
        
        if "commands" in contributes:
            permissions.append(Permission(
                name="commands",
                description=f"Registers {len(contributes['commands'])} commands",
                is_dangerous=False,
            ))
        
        return permissions
    
    def _summarize_contributes(self, contributes: Dict[str, Any]) -> Dict[str, Any]:
        """Summarize the contributes section."""
        summary = {}
        
        for key, value in contributes.items():
            if isinstance(value, list):
                summary[key] = len(value)
            elif isinstance(value, dict):
                summary[key] = len(value)
            else:
                summary[key] = 1
        
        return summary

File: pkg/detectors/vscode.go
package detectors

import (
	"crypto/sha256"
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"strings"
	"time"

	"github.com/securient/ideviewer-oss/pkg/scanner"
)

// Dangerous activation-event permissions (matches Python exactly).
var dangerousPermissions = map[string]string{
	"onFileSystem":            "Full file system access",
	"onStartupFinished":       "Runs on startup",
	"onUri":                   "Can register URI handlers",
	"onAuthenticationRequest": "Authentication access",
	"onDebug":                 "Debugger access",
	"onTerminalProfile":       "Terminal access",
	"*":                       "Wildcard activation (runs for everything)",
}

var dangerousCapabilities = map[string]string{
	"untrustedWorkspaces.supported": "Can run in untrusted workspaces",
	"virtualWorkspaces.supported":   "Virtual workspace access",
}

type vscodeVariant struct {
	ideType         scanner.IDEType
	name            string
	executables     map[string][]string
	extensionsPaths map[string][]string
	configPaths     map[string][]string
	processNames    []string
}

var vscodeVariants = []vscodeVariant{
	{
		ideType: scanner.IDETypeVSCode,
		name:    "Visual Studio Code",
		executables: map[string][]string{
			"darwin": {
				"/Applications/Visual Studio Code.app/Contents/Resources/app/bin/code",
				"/usr/local/bin/code",
			},
			"linux": {
				"/usr/bin/code",
				"/usr/share/code/bin/code",
				"/snap/bin/code",
			},
			"windows": {
				`%LOCALAPPDATA%\Programs\Microsoft VS Code\Code.exe`,
				`%ProgramFiles%\Microsoft VS Code\Code.exe`,
			},
		},
		extensionsPaths: map[string][]string{
			"darwin":  {"~/.vscode/extensions"},
			"linux":   {"~/.vscode/extensions"},
			"windows": {`%USERPROFILE%\.vscode\extensions`},
		},
		configPaths: map[string][]string{
			"darwin":  {"~/Library/Application Support/Code"},
			"linux":   {"~/.config/Code"},
			"windows": {`%APPDATA%\Code`},
		},
		processNames: []string{"Code", "code", "Code.exe"},
	},
	{
		ideType: scanner.IDETypeCursor,
		name:    "Cursor",
		executables: map[string][]string{
			"darwin": {
				"/Applications/Cursor.app/Contents/Resources/app/bin/cursor",
				"/usr/local/bin/cursor",
			},
			"linux": {
				"/usr/bin/cursor",
				"/opt/Cursor/cursor",
				"~/.local/bin/cursor",
			},
			"windows": {
				`%LOCALAPPDATA%\Programs\Cursor\Cursor.exe`,
				`%LOCALAPPDATA%\cursor\Cursor.exe`,
			},
		},
		extensionsPaths: map[string][]string{
			"darwin":  {"~/.cursor/extensions"},
			"linux":   {"~/.cursor/extensions"},
			"windows": {`%USERPROFILE%\.cursor\extensions`},
		},
		configPaths: map[string][]string{
			"darwin":  {"~/Library/Application Support/Cursor"},
			"linux":   {"~/.config/Cursor"},
			"windows": {`%APPDATA%\Cursor`},
		},
		processNames: []string{"Cursor", "cursor", "Cursor.exe"},
	},
	{
		ideType: scanner.IDETypeVSCodium,
		name:    "VSCodium",
		executables: map[string][]string{
			"darwin": {
				"/Applications/VSCodium.app/Contents/Resources/app/bin/codium",
				"/usr/local/bin/codium",
			},
			"linux": {
				"/usr/bin/codium",
				"/snap/bin/codium",
			},
			"windows": {
				`%LOCALAPPDATA%\Programs\VSCodium\VSCodium.exe`,
				`%ProgramFiles%\VSCodium\VSCodium.exe`,
			},
		},
		extensionsPaths: map[string][]string{
			"darwin":  {"~/.vscode-oss/extensions"},
			"linux":   {"~/.vscode-oss/extensions"},
			"windows": {`%USERPROFILE%\.vscode-oss\extensions`},
		},
		configPaths: map[string][]string{
			"darwin":  {"~/Library/Application Support/VSCodium"},
			"linux":   {"~/.config/VSCodium"},
			"windows": {`%APPDATA%\VSCodium`},
		},
		processNames: []string{"VSCodium", "codium", "VSCodium.exe"},
	},
}

// VSCodeDetector detects VS Code, Cursor, and VSCodium.
type VSCodeDetector struct{}

func (d *VSCodeDetector) Name() string { return "vscode" }

func (d *VSCodeDetector) Detect() ([]scanner.IDE, error) {
	var ides []scanner.IDE
	plat := PlatformKey()
	for _, v := range vscodeVariants {
		ide, ok := detectVariant(v, plat)
		if ok {
			ide.Extensions = parseVSCodeExtensions(ide.ExtensionsPath, v.ideType)
			ides = append(ides, ide)
		}
	}
	return ides, nil
}

func detectVariant(v vscodeVariant, plat string) (scanner.IDE, bool) {
	var installPath, extensionsPath, configPath string

	for _, p := range v.executables[plat] {
		ep := ExpandPath(p)
		if info, err := os.Stat(ep); err == nil && !info.IsDir() {
			installPath = ep
			break
		}
	}
	for _, p := range v.extensionsPaths[plat] {
		ep := ExpandPath(p)
		if info, err := os.Stat(ep); err == nil && info.IsDir() {
			extensionsPath = ep
			break
		}
	}
	if installPath == "" && extensionsPath == "" {
		return scanner.IDE{}, false
	}
	for _, p := range v.configPaths[plat] {
		ep := ExpandPath(p)
		if info, err := os.Stat(ep); err == nil && info.IsDir() {
			configPath = ep
			break
		}
	}

	var version string
	if installPath != "" {
		version = GetVersion(installPath)
	}

	return scanner.IDE{
		IDEType:        v.ideType,
		Name:           v.name,
		Version:        version,
		InstallPath:    installPath,
		ConfigPath:     configPath,
		ExtensionsPath: extensionsPath,
		IsRunning:      IsProcessRunning(v.processNames...),
	}, true
}

// parseVSCodeExtensions scans the extensions directory for package.json files.
func parseVSCodeExtensions(extensionsDir string, ideType scanner.IDEType) []scanner.Extension {
	if extensionsDir == "" {
		return nil
	}
	entries, err := os.ReadDir(extensionsDir)
	if err != nil {
		return nil
	}
	var exts []scanner.Extension
	for _, entry := range entries {
		if !entry.IsDir() || strings.HasPrefix(entry.Name(), ".") {
			continue
		}
		ext, ok := parseVSCodeExtension(filepath.Join(extensionsDir, entry.Name()), ideType)
		if ok {
			exts = append(exts, ext)
		}
	}
	return exts
}

type vscodePkgJSON struct {
	Name                 string            `json:"name"`
	DisplayName          string            `json:"displayName"`
	Version              string            `json:"version"`
	Publisher            string            `json:"publisher"`
	Description          string            `json:"description"`
	Homepage             string            `json:"homepage"`
	License              string            `json:"license"`
	Repository           json.RawMessage   `json:"repository"`
	Author               json.RawMessage   `json:"author"`
	ActivationEvents     []string          `json:"activationEvents"`
	Capabilities         map[string]any    `json:"capabilities"`
	Contributes          map[string]any    `json:"contributes"`
	ExtensionDeps        []string          `json:"extensionDependencies"`
}

func parseVSCodeExtension(dir string, ideType scanner.IDEType) (scanner.Extension, bool) {
	pkgPath := filepath.Join(dir, "package.json")
	data, err := os.ReadFile(pkgPath)
	if err != nil {
		return scanner.Extension{}, false
	}
	// Compute content hash for change detection
	contentHash := fmt.Sprintf("%x", sha256.Sum256(data))

	var pkg vscodePkgJSON
	if err := json.Unmarshal(data, &pkg); err != nil {
		return scanner.Extension{}, false
	}

	folderName := filepath.Base(dir)
	extID := pkg.Name
	if extID == "" {
		extID = folderName
	}
	if pkg.Publisher != "" {
		extID = pkg.Publisher + "." + extID
	}

	// Repository can be string or object
	repo := parseRepository(pkg.Repository)

	// Author / maintainer
	maintainer := parseAuthor(pkg.Author)

	// Marketplace URL
	var marketplaceURL string
	switch ideType {
	case scanner.IDETypeVSCode:
		marketplaceURL = "https://marketplace.visualstudio.com/items?itemName=" + extID
	case scanner.IDETypeVSCodium:
		marketplaceURL = fmt.Sprintf("https://open-vsx.org/extension/%s/%s", pkg.Publisher, pkg.Name)
	}

	// Last updated from file mtime
	var lastUpdated *time.Time
	if info, err := os.Stat(pkgPath); err == nil {
		t := info.ModTime()
		lastUpdated = &t
	}

	isBuiltin := strings.Contains(strings.ToLower(dir), "ms-vscode") &&
		strings.Contains(strings.ToLower(dir), "builtin")

	displayName := pkg.DisplayName
	if displayName == "" {
		displayName = pkg.Name
		if displayName == "" {
			displayName = folderName
		}
	}

	permissions := extractVSCodePermissions(pkg)
	contributes := summarizeContributes(pkg.Contributes)

	return scanner.Extension{
		ID:               extID,
		Name:             displayName,
		Version:          pkg.Version,
		Publisher:        pkg.Publisher,
		Maintainer:       maintainer,
		Description:      pkg.Description,
		Homepage:         pkg.Homepage,
		Repository:       repo,
		License:          pkg.License,
		InstallPath:      dir,
		Permissions:      permissions,
		Contributes:      contributes,
		Dependencies:     pkg.ExtensionDeps,
		Enabled:          true,
		Builtin:          isBuiltin,
		LastUpdated:      lastUpdated,
		MarketplaceURL:   marketplaceURL,
		ActivationEvents: pkg.ActivationEvents,
		Capabilities:     pkg.Capabilities,
		ContentHash:      contentHash,
	}, true
}

func extractVSCodePermissions(pkg vscodePkgJSON) []scanner.Permission {
	var perms []scanner.Permission
	seen := make(map[string]bool)

	// Activation events
	for _, event := range pkg.ActivationEvents {
		eventType := event
		if idx := strings.IndexByte(event, ':'); idx >= 0 {
			eventType = event[:idx]
		}
		if desc, ok := dangerousPermissions[eventType]; ok && !seen[eventType] {
			seen[eventType] = true
			perms = append(perms, scanner.Permission{Name: eventType, Description: desc, IsDangerous: true})
		} else if event == "*" && !seen["*"] {
			seen["*"] = true
			perms = append(perms, scanner.Permission{Name: "*", Description: dangerousPermissions["*"], IsDangerous: true})
		}
	}

	// Capabilities
	capStr := fmt.Sprintf("%v", pkg.Capabilities)
	for capName, capDesc := range dangerousCapabilities {
		if strings.Contains(capStr, capName) && !seen[capName] {
			seen[capName] = true
			perms = append(perms, scanner.Permission{Name: capName, Description: capDesc, IsDangerous: true})
		}
	}

	// Contributes
	if pkg.Contributes != nil {
		if _, ok := pkg.Contributes["authentication"]; ok && !seen["authentication"] {
			seen["authentication"] = true
			perms = append(perms, scanner.Permission{Name: "authentication", Description: "Provides authentication providers", IsDangerous: true})
		}
		if _, ok := pkg.Contributes["terminal"]; ok && !seen["terminal"] {
			seen["terminal"] = true
			perms = append(perms, scanner.Permission{Name: "terminal", Description: "Terminal integration", IsDangerous: true})
		}
		if _, ok := pkg.Contributes["debuggers"]; ok && !seen["debuggers"] {
			seen["debuggers"] = true
			perms = append(perms, scanner.Permission{Name: "debuggers", Description: "Debugger integration", IsDangerous: false})
		}
		if _, ok := pkg.Contributes["taskDefinitions"]; ok && !seen["taskDefinitions"] {
			seen["taskDefinitions"] = true
			perms = append(perms, scanner.Permission{Name: "taskDefinitions", Description: "Can define tasks (may execute commands)", IsDangerous: true})
		}
		if cmds, ok := pkg.Contributes["commands"]; ok && !seen["commands"] {
			seen["commands"] = true
			count := 0
			if arr, ok := cmds.([]any); ok {
				count = len(arr)
			}
			perms = append(perms, scanner.Permission{Name: "commands", Description: fmt.Sprintf("Registers %d commands", count), IsDangerous: false})
		}
	}

	return perms
}

func summarizeContributes(contributes map[string]any) map[string]any {
	if contributes == nil {
		return nil
	}
	summary := make(map[string]any, len(contributes))
	for k, v := range contributes {
		switch val := v.(type) {
		case []any:
			summary[k] = len(val)
		case map[string]any:
			summary[k] = len(val)
		default:
			summary[k] = 1
		}
	}
	return summary
}

func parseRepository(raw json.RawMessage) string {
	if len(raw) == 0 {
		return ""
	}
	var s string
	if json.Unmarshal(raw, &s) == nil {
		return s
	}
	var obj struct {
		URL string `json:"url"`
	}
	if json.Unmarshal(raw, &obj) == nil {
		return obj.URL
	}
	return ""
}

func parseAuthor(raw json.RawMessage) string {
	if len(raw) == 0 {
		return ""
	}
	var s string
	if json.Unmarshal(raw, &s) == nil {
		return s
	}
	var obj struct {
		Name string `json:"name"`
	}
	if json.Unmarshal(raw, &obj) == nil {
		return obj.Name
	}
	return ""
}
